Implemented javaHashUTF16LE()

achimbab opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, not needed for non-significant PRs):

...


Detailed description (optional):
This PR related to the issue https://github.com/ClickHouse/ClickHouse/issues/7643.
Java's `hashCode()` calculated a hashCode value based on a charset.
```Java
// JDK 8
public int hashCode() {
    int h = hash;
    if (h == 0 && value.length > 0) {
        hash = h = isLatin1() ? StringLatin1.hashCode(value) : StringUTF16.hashCode(value);
    }
    return h;
}
```
For compatibility with Java language, `javaHash()` function have to recognize the charsets of source string. But, ClickHouse's strings have no information about encoding.
So it is need to develop the specific function for `utf-16le`.

Commit:
Implemented javaHashUTF16LE()

alexey-milovidov committed:
Update FunctionsHashing.h

alexey-milovidov commented on dbms/src/Functions/FunctionsHashing.h:
```diff
@@ -353,6 +353,43 @@ struct JavaHashImpl
+
+        UInt32 h = 0;
+        for (size_t i = 0; i < size; i += 2)
+            h = 31 * h + static_cast<UInt32>((data[i] & 0xFF) << HI_BYTE_SHIFT | (data[i + 1] & 0xFF) << LO_BYTE_SHIFT);
```
This code is difficult to read, because it involves `char` type that is signed on x86 but unsigned on ARM,
and on x86 it involves bit shift of negative number
and cast to UInt32 with overflow.

Have you checked that it works correctly?

PS. char & 0xFF is redundant.
HI_BYTE_SHIFT and LO_BYTE_SHIFT can be inline values.
You can also use `unalignedLoad`

alexey-milovidov commented on dbms/src/Functions/FunctionsHashing.h:
```diff
@@ -353,6 +353,43 @@ struct JavaHashImpl
+        size_t size = raw_size;
+
+        // Remove Byte-order-mark(0xFFFE) for UTF-16LE
+        if (size >= 2 && data[0] == -1 && data[1] == -2)
```
You compare `char` with `int`. And char may be signed or unsigned.
Either replace `char` to `Int8` or `UInt8`
or replace -1 to `'\xFF'` and -2 to `'\xFE'`.

Commit:
Fixed type conversions in javaHashUTF16LE

achimbab commented:
@alexey-milovidov 
Thank you for your code-review. :)
I fixed code according to your recommendations and some more modifications as the below Java's `StringUTF16.hashCode()`.

```Java


class StringUTF16 {
    ...
    public static int hashCode(byte[] value) {
        int h = 0;
        int length = value.length >> 1;
        for (int i = 0; i < length; i++) {
            h = 31 * h + getChar(value, i);
        }
        return h;
    }

    public static char getChar(byte[] val, int index) {
        index <<= 1;
        return (char)(((val[index++] & 0xff) << HI_BYTE_SHIFT) |
                      ((val[index]   & 0xff) << LO_BYTE_SHIFT));
    }
}
```

After I fixed code, tested some Strings.

| String | javaHashUTF16LE |  Java's StringUTF16.hashCode() |
| ------ | ------------------- | ------------------------------------ |
|  a1가 |  138768 | 138768 |
| 가나다라마바사아자차카타파하 | -2143570108 | -2143570108 |
| FJKLDSJFIOLD_389159837589429 | 2145564783 | 2145564783  |

 If there is any other problems, I will gladly fix code.

Commit:
Removed undefined behavior of signed integer overflow.

achimbab commented:
@alexey-milovidov 
I learned a lot during code-review. 
Thank you very much.

alexey-milovidov commented:
@achimbab Is this function applicable for strings with characters not from Basic Multilingual Plane? If they are represented not as UTF-16 (surrogate pairs) in Java, then another function like `javaHashUTF32LE` will be needed.
